crates/control/src/role_assignment.rs

```diff
@@ -716,109 +768,64 @@ fn team_ball_from_seen_ball(
     }
 }
 
 fn pick_role_with_penalties(
-    own_player_number: PlayerNumber,
-    penalties: &Players<Option<Penalty>>,
-    striker_player_number: PlayerNumber,
-    optional_roles: &[Role],
+    replacement_keeper_priority: usize,
+    striker_priority: usize,
+    striker_jersey_number: u8,
+    own_jersey_number: u8,
+    offense_optional_roles: &[Role],
+    defense_optional_roles: &[Role],
+    number_of_defensive_players: usize,
 ) -> Role {
-    let mut role_assignment: Players<Option<Role>> = Players {
-        one: None,
-        two: None,
-        three: None,
-        four: None,
-        five: None,
-        six: None,
-        seven: None,
-    };
-
-    role_assignment[striker_player_number] = Some(Role::Striker);
-    let mut unassigned_robots = 6;
-
-    unassigned_robots -= penalties
-        .iter()
-        .filter(|(_player, &penalty)| penalty.is_some())
-        .count();
-
-    if unassigned_robots > 0 {
-        unassigned_robots =
-            assign_keeper_or_replacement_keeper(unassigned_robots, penalties, &mut role_assignment);
+    if replacement_keeper_priority == 0
+        || (replacement_keeper_priority == 1 && own_jersey_number > striker_jersey_number)
+    {
+        return Role::ReplacementKeeper;
     }
-
-    for &optional_role in optional_roles.iter().take(unassigned_robots) {
-        if needs_assignment(PlayerNumber::Two, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Two] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Three, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Three] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Four, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Four] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Five, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Five] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Six, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Six] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Seven, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Seven] = Some(optional_role);
+    if replacement_keeper_priority <= number_of_defensive_players {
+        defense_optional_roles
+            .get(replacement_keeper_priority - 1)
+            .copied()
+            .unwrap_or_default()
+    } else {
+        match own_jersey_number.cmp(&striker_jersey_number) {
+            Ordering::Greater => offense_optional_roles
+                .get(striker_priority)
+                .copied()
+                .unwrap_or_default(),
+            Ordering::Equal => Role::Striker,
+            Ordering::Less => offense_optional_roles
+                .get(striker_priority - 1)
+                .copied()
+                .unwrap_or_default(),
```

Comment:
I am not sure I understood everything that happens in `role_assignment.rs` and the code is horrible (not your fault). How much confidence do you have in all of this working?

Reply:
Well most of what I change is the initial role assignment in ready and the very deep down assignment for the roles based on priorities. From my testing this works (slightly differently from what we are used to) but as expected.